src/Security/Member.php

```diff
@@ -1713,63 +1719,63 @@ public function validate()
      * @param bool $write Whether to write the member afterwards
      * @return ValidationResult
      */
-    public function changePassword($password, $write = true)
+    public function changePassword($password, $write = true, $skipValidation = false)
     {
-        $this->Password = $password;
-        $valid = $this->validate();
+        /** @var PasswordHashService $hashService */
+        $hashService = Injector::inst()->get(PasswordHashService::class);
 
-        $this->extend('onBeforeChangePassword', $password, $valid);
+        // Create a password validator an check the provided password
+        $validator = static::password_validator();
+        $validationResult = $validator->validate($password, $this);
 
-        if ($valid->isValid()) {
-            $this->AutoLoginHash = null;
+        $this->extend('onBeforeChangePassword', $password, $validationResult);
 
-            $this->encryptPassword();
+        if ($skipValidation || $validationResult->isValid()) {
+            // The hash service will validate the password and set it given it's valid
+            $hashService->setForMember($password, $this, true);
 
             if ($write) {
-                $this->passwordChangesToWrite = true;
                 $this->write();
             }
         }
 
-        $this->extend('onAfterChangePassword', $password, $valid);
+        $this->extend('onAfterChangePassword', $password, $validationResult);
+
+        return $validationResult;
+    }
+
+    /**
+     * Catch updates to `Password` on a Member object
+     *
+     * @param string password
+     * @param bool $suppressNotice
+     */
+    public function setPassword($password, $suppressNotice = false)
+    {
+        // Warn against this API and assume plaintext that needs to be hashed
+//        if (!$suppressNotice) {
+//            user_error(
+//                'Setting a Password on a member should no longer be plaintext. To update a password for a member use ' .
+//                    'the changePassword method or use PasswordHashService instead',
+//                E_USER_NOTICE
+//            );
+//        }
 
-        return $valid;
+        $this->changePassword($password, false, true);
```

Comment:
This bit could be improved - I think. I've altered the functionality so that passwords are never set _unhashed_ on a Member object. This is capturing calls to `$member->Password = 'thing'`. I'll at least like to add in a deprecation notice - but I need to update places that set a password to use `PasswordHashService::setForMember` first. The challenge here is `Field::saveInto`...